kyo-prelude/shared/src/main/scala/kyo/Stream.scala

```diff
@@ -328,25 +329,52 @@ final case class Stream[V, -S](v: Ack < (Emit[Chunk[V]] & S)):
 end Stream
 
 object Stream:
+    @nowarn("msg=anonymous")
+    inline def apply[V, S](v: => Ack < (Emit[Chunk[V]] & S)): Stream[V, S] =
+        new Stream[V, S]:
+            def emit: Ack < (Emit[Chunk[V]] & S) = v
 
     private val _empty           = Stream(Stop)
     def empty[V]: Stream[V, Any] = _empty.asInstanceOf[Stream[V, Any]]
 
-    def init[V, S](seq: Seq[V] < S)(using tag: Tag[Emit[Chunk[V]]], frame: Frame): Stream[V, S] =
-        Stream[V, S](
-            seq.map { seq =>
-                val chunk: Chunk[V] = Chunk.from(seq)
-                Emit.andMap(Chunk.empty[V]) { ack =>
-                    Loop(chunk, ack) { (c, ack) =>
-                        ack match
-                            case Stop =>
-                                Loop.done(Stop)
-                            case Continue(n) =>
-                                if c.isEmpty then Loop.done(Ack.Continue())
-                                else Emit.andMap(c.take(n))(ack => Loop.continue(c.dropLeft(n), ack))
-                    }
-                }
+    def init[V, S](v: => Iterable[V] < S, chunkSize: Int = 4096)(using tag: Tag[Emit[Chunk[V]]], frame: Frame): Stream[V, S] =
+        Stream(
+            v.map {
+                case seq: Seq[V] => fromChunk(Chunk.from(seq), chunkSize)
+                case iterable    => fromIterable(iterable, chunkSize)
             }
         )
 
+    private def fromChunk[V, S](chunk: Chunk[V], chunkSize: Int)(using tag: Tag[Emit[Chunk[V]]], frame: Frame) =
+        Emit.andMap(Chunk.empty[V]) { ack =>
+            Loop(chunk, ack) { (c, ack) =>
+                ack match
+                    case Stop =>
+                        Loop.done(Stop)
+                    case Continue(n) =>
+                        if c.isEmpty then Loop.done(Ack.Continue())
+                        else
+                            val i = n min chunkSize
+                            Emit.andMap(c.take(i))(ack => Loop.continue(c.dropLeft(i), ack))
+            }
+        }
+
+    private def fromIterable[V, S](iterable: Iterable[V], chunkSize: Int)(using tag: Tag[Emit[Chunk[V]]], frame: Frame) =
+        Emit.andMap(Chunk.empty[V]) { ack =>
+            val iterator = iterable.iterator
```

Comment:
I'm not sure about this since using the iterator is a side effect

Reply:
I removed it for now. Let's see if it's necessary later.